### ISSUE TITLE
Formatting instability in `async` call chain

### ISSUE THREAD

**Issue by roshanjrajan-zip** [astral-sh/ruff#8644]
I don't know if this is expected but I am getting an error when formatting the following code. I would expect Ruff to come to a solution and not have differences when running with --no-cache or not.

Ruff Settings: Nothing specific for format but have line-length set to 88.
Ruff Version: v0.1.4/0.1.5

Steps to reproduce:
1) Create file called test.py with the following code
```python
# test.py
def func():
    test_data = await (
        Stream.from_async(async_data.data())
        .flat_map_async(lambda data: data.async_data())
        .map(lambda data: data.data())
        .filter_async(is_valid_data)
        .to_list()
    )
```

2) Call `ruff format test.py --isolated `. Can call this repeatedly and it stays the same as the following.
```python
def func():
    test_data = await Stream.from_async(async_data.data()).flat_map_async(
        lambda data: data.async_data()
    ).map(lambda data: data.data()).filter_async(is_valid_data).to_list()
```

3) Now if I run this code in CI where it runs the github action on v0.1.4 with `format check` it fails. When I call `ruff format test.py --isolated --no-cache` it formats the code to the following.
```python
def func():
    test_data = (
        await Stream.from_async(async_data.data())
        .flat_map_async(lambda data: data.async_data())
        .map(lambda data: data.data())
        .filter_async(is_valid_data)
        .to_list()
    )
```

This is stable and all calls to ruff format w/o the --no-cache flag do not reformat the file. This also allows it to pass in CI with the format check. Any help/pointers on how to make sure this is stable is greatly appreciated!

**Comment by charliermarsh** [comment]
I think the issue is that there's an instability in the formatting:

```python
# test.py
def func():
    test_data = await (
        Stream.from_async(async_data.data())
        .flat_map_async(lambda data: data.async_data())
        .map(lambda data: data.data())
        .filter_async(is_valid_data)
        .to_list()
    )

# Format once...
def func():
    test_data = await Stream.from_async(async_data.data()).flat_map_async(
        lambda data: data.async_data()
    ).map(lambda data: data.data()).filter_async(is_valid_data).to_list()

# Format twice...
def func():
    test_data = (
        await Stream.from_async(async_data.data())
        .flat_map_async(lambda data: data.async_data())
        .map(lambda data: data.data())
        .filter_async(is_valid_data)
        .to_list()
    )
```

It's just a bug in the formatter. Will need to ensure it's fixed for the next release. Thanks!

(If you want a workaround, I'd suggest just applying that final formatting, which Ruff will retain.)

**Comment by roshanjrajan-zip** [comment]
Will do and thanks for looking into this! Really appreciate the great work you are doing! 🔥

**Comment by charliermarsh** [comment]
Closed by https://github.com/astral-sh/ruff/pull/8676.
